### PR TITLE
feat: Adds support for IUM

### PR DESCRIPTION
**Pull request by jpvajda**
## Proposed changes

- Adds support for [inject user message](https://developers.deepgram.com/docs/voice-agent-inject-user-message)
- Adds unit test to test the code
- Adds a new daily test to test the Agent API
-  updates .gitignore to ignore response_data output

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update or tests (if none of the other choices apply)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I have lint'ed all of my code using repo standards
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Summary by CodeRabbit

* **New Features**
  * Added the ability to inject user messages into the agent via the websocket client, enabling direct user-to-agent text interactions.
* **Bug Fixes**
  * Updated the ignore list to ensure test artifacts are properly excluded.
* **Tests**
  * Introduced unit tests to verify user message injection and serialization functionality.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210489778526221

### CODE REVIEW

**Comment by jpvajda** [comment]
We'll release this along with Speak Fallback which will come in another PR.

File: .gitignore [hunk 1]
```diff
@@ -28,3 +28,6 @@ output_*.wav
 # test artifacts
 output.wav
 test.mp3
+
+# test data
+tests/response_data
```
Comment (lukeocodes) [inline .gitignore]: What is response data, is this the snapshot data?

File: deepgram/clients/agent/v1/websocket/async_client.py [hunk 2]
```diff
@@ -619,6 +620,54 @@ async def keep_alive(self) -> bool:
...
+
+        if not ret:
+            self._logger.error("send_function_call_response failed")
+            self._logger.spam("AsyncAgentWebSocketClient.send_function_call_response LEAVE")
+            return False
+
+        self._logger.notice("send_function_call_response succeeded")
+        self._logger.spam("AsyncAgentWebSocketClient.send_function_call_response LEAVE")
+
+        return True
```
Comment (lukeocodes) [inline deepgram/clients/agent/v1/websocket/async_client.py]: Why was this added?

File: deepgram/clients/agent/v1/websocket/client.py [hunk 3]
```diff
@@ -615,6 +616,54 @@ def keep_alive(self) -> bool:
...
+
+        if not ret:
+            self._logger.error("send_function_call_response failed")
+            self._logger.spam("AgentWebSocketClient.send_function_call_response LEAVE")
+            return False
+
+        self._logger.notice("send_function_call_response succeeded")
+        self._logger.spam("AgentWebSocketClient.send_function_call_response LEAVE")
+
+        return True
```
Comment (lukeocodes) [inline deepgram/clients/agent/v1/websocket/client.py]: It's confusing why the IUM feature has this function

File: tests/unit_test/test_unit_agent_inject_user_message.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+class TestAgentInjectUserMessage:
+    """Focused unit tests for inject_user_message functionality"""
+
+    def test_inject_user_message_options_serialization(self):
+        """Test JSON serialization is correct"""
+        options = InjectUserMessageOptions(content="Test message")
+        result = json.loads(str(options))
+        expected = {"type": "InjectUserMessage", "content": "Test message"}
+        assert result == expected
+
```
Comment (lukeocodes) [inline tests/unit_test/test_unit_agent_inject_user_message.py:19]: I believe this is the only additional test this work needed...

File: .gitignore (hunk 1 above)
Comment (jpvajda) [reply .gitignore]: Yes, it's something I think David setup, I wasn't sure if this output should  be pushed up to repo given it's generated when you run tests locally.

If we should be update these results just let me know.

File: tests/unit_test/test_unit_agent_inject_user_message.py (hunk 4 above)
Comment (jpvajda) [reply tests/unit_test/test_unit_agent_inject_user_message.py:19]: I agree. it's the unit test.  I added a daily test since one didn't exist for Agent. Should I just remove the daily test from this PR? 

The new daily test works.

File: deepgram/clients/agent/v1/websocket/client.py (hunk 3 above)
Comment (jpvajda) [reply deepgram/clients/agent/v1/websocket/client.py]: Let me look into this.

File: deepgram/clients/agent/v1/websocket/async_client.py (hunk 2 above)
Comment (jpvajda) [reply deepgram/clients/agent/v1/websocket/async_client.py]: Let me look into this.

File: deepgram/clients/agent/v1/websocket/client.py (hunk 3 above)
Comment (jpvajda) [reply deepgram/clients/agent/v1/websocket/client.py]: This was incorrectly added by Cursor, I missed it,  That's my bad.

Removing it in the next commit

File: deepgram/clients/agent/v1/websocket/async_client.py (hunk 2 above)
Comment (jpvajda) [reply deepgram/clients/agent/v1/websocket/async_client.py]: This was incorrectly added by Cursor, I missed it,  That's my bad. Removing it in the next commit.

File: tests/unit_test/test_unit_agent_inject_user_message.py (hunk 4 above)
Comment (jpvajda) [reply tests/unit_test/test_unit_agent_inject_user_message.py:19]: Decision: I'll remove the new daily test. I'll saved the code an can consider adding it later.  will fix in next commit.

File: .gitignore (hunk 1 above)
Comment (jpvajda) [reply .gitignore]: removing this as well. I'll assume this is what we want. next commit will take care of it.

**Comment by jpvajda** [comment]
@lukeocodes  all review comments are resolved.

**Review by lukeocodes — APPROVED** [review]
why is this so complex wtf

approved!!

**Comment by jpvajda** [comment]
I built a simple test app locally and this looks good. ✅ 


🎉 **Excellent! All tests passed successfully!**

The test output shows that your inject_user_message feature is working perfectly:

## Test Results Summary:

✅ **Message Serialization (6/6 tests passed)**
- Correctly serializes all message types including special characters, empty messages, and Unicode
- Proper JSON structure with `type: "InjectUserMessage"` and `content` field

✅ **Inject User Message Method (4/4 tests passed)**  
- Method returns `True` when successful
- Correctly calls the underlying `send()` method with properly serialized message
- Works with various message lengths and content

✅ **Invalid Input Handling (5/5 tests passed)**
- Properly rejects invalid input types (strings, integers, None, lists, dicts)
- Returns `False` for all invalid inputs
- Shows proper error logging: "options must be of type InjectUserMessageOptions"

✅ **WebSocket Send Failure (1/1 test passed)**
- Correctly handles failure scenarios when websocket send fails
- Returns `False` when underlying transport fails
- Shows proper error logging: "inject_user_message failed"


**Future Steps**: The inject_user_message feature is confirmed to be working properly beyond your unit tests. You can now confidently use this feature knowing it handles all the edge cases properly.

The test demonstrated excellent error handling, proper serialization, and robust failure scenarios - exactly what you'd want to see in a production feature!
